File: cli/backend/book.js
const arabicTrans = require("./transliteration");

function Book(id, name, url, index, publisher, authors, cover, type, thumb, access, file_md5, file_id, latest_file_id, size, header, key, bookPath) {
    if (typeof id === "object") {
        Object.assign(this, id);
        return this;
    }

    if (!(this instanceof Book)) {
        if (typeof id === "object") {
            return new Book(id);
        } else {
            return new Book(id, name, url, index, publisher, authors, cover, type, thumb, access, file_md5, file_id, latest_file_id, size, header, key, bookPath);
        }
    }

    if (!["pdf", "epub", "mp3"].includes(type)) {
        type = "-";
    }

    this.id = id;
    this.value = index;
    this.title = name;
    this.url = url; // this is a sample now
    this.type = type;
    this.short = this.id;
    this.publisher = publisher;
    this.authors = authors || [];
    this.cover = cover;
    this.thumb = thumb;
    this.bookPath = bookPath;
    this.name = this.id + { pdf: "🟥", epub: "📗", mp3: "🔉 mp3", "-": "" }[this.type] + ": " + arabicTrans(this.title) + "  (" + this.title + ")";
    //new items for future use
    this.access = access || 0;
    this.file_md5 = file_md5 || "";
    this.header = header || "";
    this.key = key ||  [115, -36, 110, -93, 78, -22, 63, -71, 97, -126, 86, 66, -36, 46, 13, -96];
    this.file_id = file_id || id || 0;
    this.latest_file_id = latest_file_id || file_id || id || 0;
    this.size = size || 0;
}

Book.prototype.urlFilename = function urlFilename() {
    return this.url.split("/").pop();
};
Book.prototype.setInfo = function setInfo(item, value) {
    return (this[item] = value);
};

module.exports = Book;

File: src-tauri/src/backend/book.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Book {
    pub id: String,
    pub title: String,
    pub url: String,
    #[serde(rename = "type")]
    pub book_type: String,
    pub publisher: String,
    pub authors: Vec<String>,
    pub cover: Option<String>,
    pub thumb: Option<String>,
    pub book_path: Option<String>,
    pub name: String,
    pub access: bool,
    pub file_md5: String,
    pub header: String,
    pub key: Vec<i32>,
    pub file_id: String,
    pub latest_file_id: String,
    pub size: u64,
    pub downloaded_at: Option<u64>,
}

impl Default for Book {
    fn default() -> Self {
        Book {
            id: "".to_string(),
            title: "".to_string(),
            url: "".to_string(),
            book_type: "".to_string(),
            publisher: "".to_string(),
            authors: Vec::new(),
            cover: None,
            thumb: None,
            book_path: None,
            name: "".to_string(),
            access: false,
            file_md5: "".to_string(),
            header: "".to_string(),
            key: Vec::from([
                115, -36, 110, -93, 78, -22, 63, -71, 97, -126, 86, 66, -36, 46, 13, -96,
            ]),
            file_id: "".to_string(),
            latest_file_id: "".to_string(),
            size: 0,
            downloaded_at: None,
        }
    }
}
